pkg/trace/context.go

```diff
@@ -0,0 +1,134 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package trace
+
+import (
+	"fmt"
+	"sync/atomic"
+	"time"
+
+	"github.com/Sirupsen/logrus"
+
+	"golang.org/x/net/context"
+)
+
+const OpTraceKey = "traceKey"
+
+// monotonic counter which inrements on Start()/NewOperation()
+var opCount uint64
+
+type Operation struct {
+	OpFrame
+	context.Context
+	parent *Operation
+}
+
+type OpFrame struct {
+	Tr    Trace
+	OpNum uint64
+}
+
+// Start a Operation context
+func NewOperation(msg string) Operation {
+
+	o := Operation{
+		Context: context.Background(),
+		OpFrame: OpFrame{
+			// start the trace
+			Tr:    *newTrace(msg),
+			OpNum: atomic.AddUint64(&opCount, 1),
+		},
+	}
+
+	o.Debugf(o.Tr.beginHdr())
+	return o
+}
+
+// Creates a header string to be printed.
+func (o *Operation) header() string {
+	if Logger.Level >= logrus.DebugLevel {
+		return fmt.Sprintf("op=%d (delta:%s)", o.OpNum, o.Tr.delta())
+	} else {
+		return fmt.Sprintf("op=%d", o.OpNum)
+	}
+}
+
+// Err returns a non-nil error value after Done is closed.  Err returns
+// Canceled if the context was canceled or DeadlineExceeded if the
+// context's deadline passed.  No other values for Err are defined.
+// After Done is closed, successive calls to Err return the same value.
+func (o Operation) Err() error {
+	if err := o.Context.Err(); err != nil {
+		o.Errorf("%s: %s error: %s", o.Tr.endHdr(), o.Tr.msg, err.Error())
+
+		// Deceptively recursive.  Call this routine on the parent (which will traverse up the stack).
+		if o.parent != nil {
+			o.parent.Err()
+		}
+
+		return err
+	}
+
+	return nil
+}
+
+// Infof logs an info message prepended with the operation, the delta since the operation was created, and the specified message
+func (o Operation) Infof(format string, args ...interface{}) {
+	Logger.Infof("%s: %s", o.header(), fmt.Sprintf(format, args...))
+}
+
+// Debugf logs a debug message prepended with the operation, the delta since the operation was created, and the specified message
+func (o Operation) Debugf(format string, args ...interface{}) {
+	Logger.Debugf("%s: %s", o.header(), fmt.Sprintf(format, args...))
+}
+
+// Errorf logs an error message prepended with the operation, the delta since the operation was created, and the specified message
+func (o Operation) Errorf(format string, args ...interface{}) {
+	Logger.Errorf("%s: %s", o.header(), fmt.Sprintf(format, args...))
+}
+
+// Pass the trace as a param because newTrace will grab the frame before this
+// one.  If we call newTrace in this function, all of our error stacks will be
+// from this package (which is useless)
+func withContext(parent Operation, tr Trace, ctx context.Context) Operation {
```

Comment:
could you not just have a newTrace impl that uses 3 as the skip value for use here?